projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewStoreReader.java

```diff
@@ -396,22 +397,17 @@ List<Map<String, Set<ValueDTO>>> retrieveJoinTableRows(
                 Stream.of("id", "label"),
                 joinView.include.stream()
                         .filter(column -> !valueSetProperties.contains(column)))
+                .distinct()
                 .collect(Collectors.toList());
 
-        var query = connection.prepareStatement(
-                "select " + projectionColumns.stream().map(String::toLowerCase).collect(Collectors.joining(", ")) +
-                        " from " + joinedTable.name + " j " +
-                        " where exists (" +
-                        "   select * from " + joinTable.name + " jt " +
-                        "   where jt." + idColumn(joinView.view).name + " = j.id" +
-                        "   and jt." + idColumn(view).name + " = ? )" +
-                        "limit " + String.valueOf(MAX_JOIN_ITEMS)
-        );
-        query.setString(1, id);
+        var query = getQuery(view, joinView, joinedTable, joinTable, projectionColumns, ids);
         var result = query.executeQuery();
-        List<Map<String, Set<ValueDTO>>> rows = new ArrayList<>();
+
+        var rows = new ViewRowCollection();
+
         while (result.next()) {
-            Map<String, Set<ValueDTO>> row = new HashMap<>();
+            var row = new ViewRow();
```

Comment:
for the sake of readability I would extract the row building to a separate method